### PR TITLE
PD-50 Define Fundamental Data Structure for Diagram

### PR DESCRIPTION
**Pull request by crapper**

### CODE REVIEW

File: src/diagram/Diagram.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  /**
+   * a method that moves the field from the `from` index to the `to` index
+   *
+   * @param from the index of the field to be moved
+   * @param to   the index of the field to be moved to
+   */
+  moveField(from: number, to: number) {
+    let field: Field = this._fields.splice(from, 1)[0];
```
Comment (Jerrylum) [inline src/diagram/Diagram.ts:126]: use `const`

File: src/diagram/Diagram.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
+import { Config } from "dompurify";
+import { Field } from "./Field";
+import { Configuration } from "../config/Configuration";
+import { BooleanOption, EnumOption, RangeOption } from "../config/Option";
+
+interface FieldPair {
+  name: string;
```
Comment (Jerrylum) [inline src/diagram/Diagram.ts:7]: Use `readonly`? cuz it is for class `Memeto` and I want to make sure the values in the Memento class are unmodifiable.

File: src/diagram/Diagram.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
+import { Config } from "dompurify";
+import { Field } from "./Field";
+import { Configuration } from "../config/Configuration";
+import { BooleanOption, EnumOption, RangeOption } from "../config/Option";
+
+interface FieldPair {
```
Comment (Jerrylum) [inline src/diagram/Diagram.ts:6]: Change the interface name to `MementoFieldPair`

File: src/diagram/Diagram.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  name: string;
+  length: number;
+}
+let fields: ReadonlyArray<FieldPair> = [];
```
Comment (Jerrylum) [inline src/diagram/Diagram.ts:10]: What's this?

File: src/diagram/Diagram.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+ */
+export class Diagram {
+  /**
+   * the list of fields that the diagram holds
+   */
+  private _fields: Field[] = [];
+  /**
+   * the configuration of the diagram
+   */
+  private config: Configuration;
```
Comment (Jerrylum) [inline src/diagram/Diagram.ts:35]: Use readonly and remove the function `getConfig`

**Review by Jerrylum — CHANGES_REQUESTED** [review]

File: src/diagram/Diagram.ts (hunk 4 above)
Comment (crapper) [reply src/diagram/Diagram.ts:10]: removed.

File: src/diagram/Diagram.ts (hunk 3 above)
Comment (crapper) [reply src/diagram/Diagram.ts:6]: updated.

File: src/diagram/Diagram.ts (hunk 2 above)
Comment (crapper) [reply src/diagram/Diagram.ts:7]: updated.

File: src/diagram/Diagram.ts (hunk 1 above)
Comment (crapper) [reply src/diagram/Diagram.ts:126]: resolved.

**Review by Jerrylum — APPROVED** [review]
Note that you also modified the Configuration constructor, but I will give it a pass.

LGTM

**Pull request merged**